plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/Exception.hs

```diff
@@ -66,13 +66,30 @@ data MachineError fun
     deriving stock (Show, Eq, Functor, Generic)
     deriving anyclass (NFData)
 
--- | The type of errors (all of them) which can occur during evaluation
--- (some are used-caused, some are internal).
-data EvaluationError user internal
-    = InternalEvaluationError !internal
-      -- ^ Indicates bugs.
-    | UserEvaluationError !user
-      -- ^ Indicates user errors.
+{- | The type of errors that can occur during evaluation. There are two kinds of errors:
+
+1. Operational ones -- these are errors that are indicative of the _logic_ of the program being
+   wrong. For example, 'Error' was executed or 'tailList' was applied to an empty list.
+2. Structural ones -- these are errors that are indicative of the _structure_ of the program being
+   wrong. For example, a free variable was encountered during evaluation, or a non-function was
+   applied to an argument
+
+On the chain both of these are just regular failures and we don't distinguish between them there:
+if a script fails, it fails, it doesn't matter what the reason was. However in the tests it does
+matter why the failure occurred: a structural error may indicate that the test was written
+incorrectly while an operational error may be entirely expected.
+
+In other words, operational errors are regular runtime errors and structural errors are \"runtime
+type errors\". Which means that evaluating an (erased) well-typed program should never produce a
+structural error, only an operational one. This creates a sort of \"runtime type system\" for UPLC
+and it would be great to stick to it and enforce in tests etc, but we currently don't. For example,
+a built-in function expecting a list but getting something else should throw a structural error,
+but currently it'll throw an operational one. This is something that we plan to improve upon in
+future.
+-}
```

Comment:
This is the only important part of this PR.

Reply:
How about `data RuntimeError = RuntimeEvalError | RuntimeTypeError` ?

Reply:
I'd like to keep it intentionally separate from the whole types realm, we're talking about something entirely different here. `RuntimeTypeError` suggests that it's something about types but

1. we don't have types in UPLC
2. a missing case branch isn't a type error by any measure, yet we consider it to be a structural error. Is that a good idea? Dunno, but that's what we do

So let's just introduce new terminology for this thing that doesn't seem to match any typery exactly.